fix(common): Don't run preconnect assertion on the server.

JeanMeche opened the pull request:
The `window` global is patched by domino on the server but the value of `window.location.href` isn't a valid base.

Before this change `getUrl()` would throw when running in devmode on the server.

Fixes #56207

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -102,7 +104,7 @@ export class PreconnectLinkChecker {
    */
   assertPreconnect(rewrittenSrc: string, originalNgSrc: string): void {
-    if (!this.window) return;
+    if (!this.window || this.isServer) return;
```
I'm wondering if we should just check if the current platform is server and exit in this case:
```suggestion
    if (this.isServer) return;
```
I think that was the intention of the `window`-based check.

JeanMeche replied:
I kept it because `this.window ` is typed as `Window | null`

AndrewKushnir replied:
Thanks for additional information. In this case we can move this check to the constructor:
```
    if (!this.isServer && typeof win !== 'undefined') {
      this.window = win;
    }
```
So we set the window reference only when we are not on the server. What do you think?

alan-agius4 replied:
@JeanMeche, I understand you didn't originally work on this part, but there's an issue with how we're obtaining the base href, and it also highlights a couple of other issues.

1. We're currently using the Window object, which is incorrect. Instead, we should be using the `PlatformLocation` service. This service will return the correct base when running on the server. When an absolute URL is provided, the `PlatformLocation` service handles it correctly.
2. When a relative URL is provided in the server config, we're currently setting the base href to `/`, which is not valid, as you pointed out. This setting lacks a protocol and makes the `href` invalid. https://github.com/angular/angular/blob/101edda0184b458d331163b7e04280c3b8b48bfc/packages/platform-server/src/domino_adapter.ts#L24. 

Maybe the `getURL` function should be updated to use a `base` instead of a Window object:

```ts
export function getUrl(src: string, base: string): URL
```

This approach is essentially equivalent to using `new URL()`, because if `src` is absolute, `new URL(src)` yields the same result as `new URL(src, win.location.href)`.

AndrewKushnir replied:
@alan-agius4 quick question: the change is in this PR should make this service a noop on the server (only be invoked in a browser). Would you recommend changing the mentioned logic anyway?

JeanMeche replied:
I agree that accessing the window global isn't great. I've refactored to use `PlatformLocation`.

alan-agius4 commented on packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:
```diff
@@ -42,14 +45,13 @@ export class LCPImageObserver implements OnDestroy {
-  private window: Window | null = null;
   private observer: PerformanceObserver | null = null;
+  private platformLocation = inject(PlatformLocation);
+  private readonly isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
```
NIT: Move this into the constructor as it is not used anywhere else. Since properties are not minified, this approach will also help to optimize the code.

alan-agius4 commented on packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:
```diff
@@ -98,20 +100,20 @@ export class LCPImageObserver implements OnDestroy {
 
   updateImage(originalSrc: string, newSrc: string) {
-    const originalUrl = getUrl(originalSrc, this.window!).href;
+    const originalUrl = getUrl(originalSrc, this.platformLocation.href).href;
```
Inside of accessing `this.platformLocation.href` multiple times in the same method, store the value in a `const`. This will save a few bytes.

alan-agius4 commented on packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:
```diff
@@ -42,14 +45,13 @@ export class LCPImageObserver implements OnDestroy {
 
-  private window: Window | null = null;
   private observer: PerformanceObserver | null = null;
+  private platformLocation = inject(PlatformLocation);
```
You can also change this to the below to save some extra bytes.

```suggestion
  private baseUrl = inject(PlatformLocation).href;
```

alan-agius4 reviewed (commented):
LGTM, couple of minor nits.

AndrewKushnir commented:
[Presubmit](http://test/OCL:646510455:BASE:646510494:1719333790508:116ec2fe).

AndrewKushnir commented:
**Caretaker note**: presubmit is "green" for the changes in this PR, this PR is ready for merge.

thePunderWoman commented:
This PR was merged into the repository by commit 39e48ce6757ac2dc4157ab69acade0ff1cd2208e.

The changes were merged into the following branches: main, 18.0.x

AndrewKushnir commented:
The change will be reverted via https://github.com/angular/angular/pull/56740, since there are some breakages in Google's codebase. The error looks like this:

```
TypeError: Failed to construct 'URL': Invalid base URL

    at getUrl (packages/common/src/directives/ng_optimized_image/url.ts:12
    at LCPImageObserver.registerImage (packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:103
    at packages/common/src/directives/ng_optimized_image/ng_optimized_image.ts:438
   ...
```

I think we may need to split this PR into 2:

- one that would add a flag to stop running some code on the server (low risk)
- additional refactoring to use platform location, etc (higher risk, would require a TGP)

JeanMeche commented:
Back to review. 
This is now the minimal fix by checking for the platform.

AndrewKushnir commented:
> This is now the minimal fix by checking for the platform.

@JeanMeche I still see changes related to platform location in the diff. Could you please check if you've pushed the most recent version of the change?

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:
```diff
@@ -98,20 +97,20 @@ export class LCPImageObserver implements OnDestroy {
+    this.images.delete(getUrl(rewrittenSrc, window).href);
   }
 
   updateImage(originalSrc: string, newSrc: string) {
```
It'd be great to add an exit condition into this function as well:
```suggestion
  updateImage(originalSrc: string, newSrc: string) {
    if (!this.observer) return;
```

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts:
```diff
@@ -42,14 +43,12 @@ export class LCPImageObserver implements OnDestroy {
   constructor() {
+    const isBrowser = isPlatformBrowser(inject(PLATFORM_ID));
     assertDevMode('LCP checker');
-    const win = inject(DOCUMENT).defaultView;
-    if (typeof win !== 'undefined' && typeof PerformanceObserver !== 'undefined') {
```
I think it'd be great to keep the logic that uses `DOCUMENT` as is, since this token can be overridden (e.g. for testing purposes).

AndrewKushnir reviewed (commented):
@JeanMeche thanks for addressing the feedback 👍 I think we can simplify the change more and keep the original code that works with the `DOCUMENT` token.

AndrewKushnir commented on packages/common/src/directives/ng_optimized_image/preconnect_link_checker.ts:
```diff
@@ -68,16 +70,10 @@ export class PreconnectLinkChecker {
 
   constructor() {
     assertDevMode('preconnect link checker');
-    const win = this.document.defaultView;
```
Could you please revert these changes (similar to `packages/common/src/directives/ng_optimized_image/lcp_image_observer.ts`) as well to minimize the diff?

JeanMeche committed:
fix(common): Don't run preconnect assertion on the server.

The `window` global is patched by domino on the server but the value of `window.location.href` isn't a valid base.

Before this change `getUrl()` would throw when running in devmode on the server.

Fixes #56207

AndrewKushnir commented:
[Presubmit](http://test/OCL:646510455:BASE:651840610:1720809868761:339452dd).

AndrewKushnir commented:
**Caretaker note**: the presubmit is "green", this PR is ready for merge. **Important note**: this PR was previously reverted from g3 due to some breakages introduced by extra refactoring. This version of PR contains the minimum set of changes required, so the risk of rollback is low. However, please merge and sync this PR individually (preferably early in the week as well).

atscott commented:
This PR was merged into the repository by commit 2c4613a002d7670f8377cb53eaa9aca4bfc9521f.

The changes were merged into the following branches: main, 18.1.x
